### ISSUE TITLE
Installation failed - Lua

### ISSUE THREAD

**Issue by neonRaphael** [Project-OSRM/osrm-backend#1580]
Hi,

Unfortunately I am unable to install the osrm-backend because of a problem with luabind.
I have already tried the installation on Ubuntu 14.04, 15.04, Debian Squeeze and Debian Jessie, however it didn't work on any OS.
Is there a known issue with lua 5.1 or 5.2? I read that both versions are supported, but it doesn't work for me.

```
-- The C compiler identification is GNU 4.9.2
-- The CXX compiler identification is GNU 4.9.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Building on a 64 bit system
-- Configuring OSRM in release mode
-- Performing Test LTO_AVAILABLE
-- Performing Test LTO_AVAILABLE - Success
-- Performing Test LTO_WORKS
-- Performing Test LTO_WORKS - Success
-- LTO working
-- Using gcc specific binutils for LTO.
-- Performing Test HAS_COLOR_FLAG
-- Performing Test HAS_COLOR_FLAG - Success
-- Boost version: 1.55.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   iostreams
--   program_options
--   regex
--   system
--   thread
--   unit_test_framework
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Intel TBB
-- Looking for Luabind...
-- Found Luabind: /usr/lib/libluabind.so  
-- Found Luabind: /usr/lib/libluabind.so
-- Found Lua52: /usr/lib/x86_64-linux-gnu/liblua5.2.so;/usr/lib/x86_64-linux-gnu/libm.so  
-- Looking for LuaJIT 5.2
-- Could NOT find LUAJIT (missing:  LUAJIT_LIBRARIES LUAJIT_INCLUDE_DIR) 
-- Performing Test LUABIND_WORKS
-- Performing Test LUABIND_WORKS - Failed
-- Luabind/Lua5.2 not feasible, falling back to Lua 5.1.
-- Found Lua51: /usr/lib/x86_64-linux-gnu/liblua5.2.so;/usr/lib/x86_64-linux-gnu/libm.so (found version "5.1.5") 
-- Looking for LuaJIT 5.1
-- Could NOT find LUAJIT (missing:  LUAJIT_LIBRARIES LUAJIT_INCLUDE_DIR) 
-- Performing Test LUABIND51_WORKS
-- Performing Test LUABIND51_WORKS - Failed
CMake Error at cmake/check_luabind.cmake:38 (message):
  Luabind does not work with Lua 5.1 at
  /usr/lib/x86_64-linux-gnu/liblua5.1.so, no working Luabind found
Call Stack (most recent call first):
  CMakeLists.txt:251 (include)
```

Kind regards

**Comment by daniel-j-h** [comment]
I'm building on both Debian Jessie and Ubuntu 15.04 (and 14.04) here. Make sure to get all the dependencies: I updated the Wiki a few weeks ago to include the Ubuntu 15.04 dependencies [here](https://github.com/Project-OSRM/osrm-backend/wiki/Building-on-Ubuntu#ubuntu-1504).

Which luabind version are you on? For example, on Ubuntu 15.04 the package luabind depends on liblua52. Do you install the packages as described in the Wiki using the package manager?

**Comment by neonRaphael** [comment]
I have the following luabind packages installed:
```
libluabind-dev          0.9.1+dfsg-9            amd64        luabind c++ binding for lua: static library and headers
libluabind0.9.1         0.9.1+dfsg-9            amd64        luabind c++ binding for lua: runtime library
```

I did the build exactly as described in [Building on Ubuntu] (https://github.com/Project-OSRM/osrm-backend/wiki/Building-on-Ubuntu#ubuntu-1504) and [General build instructions from source] (https://github.com/Project-OSRM/osrm-backend/wiki/Building-OSRM#general-build-instructions-from-source)

**Comment by daniel-j-h** [comment]
Could you manually verify that you can compile a small self contained example, such as [this C++ program](https://github.com/Project-OSRM/osrm-backend/blob/300d901618dd259cfba408d0e3d96aeec4f40369/cmake/check_luabind.cmake#L4)?

That is, try to compile this program with something like

    g++ -std=c++14 -Wall -Wextra -pedantic test.cc $(pkg-config --cflags --libs luabind lua5.2)


Also check https://github.com/Project-OSRM/osrm-backend/issues/1196.

**Comment by neonRaphael** [comment]
Sorry, I have no clue about that one.
```
check_luabind.cc:1:9: error: expected constructor, destructor, or type conversion before ‘(’ token
 INCLUDE (CheckCXXSourceCompiles)
         ^
check_luabind.cc:5:57: error: expected unqualified-id before ‘)’ token
 set (CMAKE_TRY_COMPILE_CONFIGURATION ${CMAKE_BUILD_TYPE})
                                                         ^
```

#1196 
> Make sure you don't have a parallel installation of version 5.1 and 5.2.
I only installed the packages listed in the wiki article about Ubuntu. Anyway, "liblua5.2-0:amd64" and "liblua5.2-dev:amd64" have also been installed.

**Comment by daniel-j-h** [comment]
Please compile only the C++ code, not the CMake script. That is, put the following into a file and compile it as described above:

```C++
#include "lua.h"
#include <luabind/luabind.hpp>

int main() {
  lua_State *myLuaState = luaL_newstate();
  luabind::open(myLuaState);
}
```

**Comment by neonRaphael** [comment]
Thanks for the code snippet.
```
virtual memory exhausted: Cannot allocate memory
```
Have established a swap file, now the build worked. Sometimes the solution is so easy.

Thank you really much for your help :)

**Comment by daniel-j-h** [comment]
Great! Have fun exploring the features and codebase :)

**Comment by cvvergara** [comment]
I tried this code snipet
```
#include "lua.h"
#include "lauxlib.h"
#include <luabind/luabind.hpp>
int main() {
    lua_State *myLuaState = luaL_newstate();
    luabind::open(myLuaState);
}
```
which is the current code tested on [check_luabind.cmake](https://github.com/Project-OSRM/osrm-backend/blob/master/cmake/check_luabind.cmake#L4)

with out the `#include "lauxlib.h"` I have no problem on this test snippet, but with it i get:
```
luatest.cc:(.text+0x9): undefined reference to `luaL_newstate()'
```

**Comment by cvvergara** [comment]
with this one:
```
extern C {
#include "lua.h"
#include "lauxlib.h"
}
#include <luabind/luabind.hpp>
int main() {
    lua_State *myLuaState = luaL_newstate();
    luabind::open(myLuaState);
}
```
with the "extern C" it gets worse, lots of compiling errors.

**Comment by daniel-j-h** [comment]
Can you provide more context? Lua version, Luabind version, etc.
Can you try with this Luabind first: https://github.com/mapbox/luabind

> undefined reference to `luaL_newstate()'

Means is a linker issue --- you're failing to link `liblua`.

**Comment by cvvergara** [comment]
probably was that my compiler does not have c++14
gcc-4.8
but one of the things I did was to edit the CMakelist and changed to use std=c++1y and it could compile

**Comment by cvvergara** [comment]
I have in pgRouting, this section on the [CMakelists ](https://github.com/pgRouting/pgrouting/blob/develop/CMakeLists.txt#L153) to test if the flag is supported, we are not using c++14, but it can give an idea of an additional test on yours.

**Comment by pirogovsky85** [comment]
Same problem, use Ubuntu 14.04.1:
`cmake .. -DCMAKE_BUILD_TYPE=Release
-- Building on a 64 bit system
-- Configuring OSRM in release mode
-- LTO working
-- Disabling LTO on GCC < 4.9.0 since it is broken, see: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=57038
-- Setting linker optimizations
-- Boost version: 1.54.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   iostreams
--   program_options
--   regex
--   system
--   thread
-- Found Intel TBB
-- TBB interface version: 7000
-- Looking for Luabind...
-- Found Luabind: /usr/local/lib/libluabind.so
-- Looking for LuaJIT 5.2
-- Found LuaJIT: /usr/lib/x86_64-linux-gnu/libluajit-5.1.so
-- Performing Test LUABIND_WORKS
-- Performing Test LUABIND_WORKS - Failed
-- Luabind/Lua5.2 not feasible, falling back to Lua 5.1.
-- Looking for LuaJIT 5.1
-- Found LuaJIT: /usr/lib/x86_64-linux-gnu/libluajit-5.1.so
-- Performing Test LUABIND51_WORKS
-- Performing Test LUABIND51_WORKS - Failed
CMake Error at cmake/check_luabind.cmake:45 (message):
  Luabind does not work with Lua 5.1 at
  /usr/lib/x86_64-linux-gnu/liblua5.1.so, no working Luabind found
Call Stack (most recent call first):
  CMakeLists.txt:284 (include)`

**Comment by daniel-j-h** [comment]
Follow-up since it seems like we forgot about this conversation: master now has a fix for gcc 4.8 and std=c++14 for now. We don't plan on supporting gcc 4.8 in the future though.

**Comment by cvvergara** [comment]
@daniel-j-h 
Thanks, that will make some users not need to upgrade their compiler immediately.
Actually, after stumbling on this when trying to use osrm, now I have different g++ compilers on my computer and switch back & forth with
```
sudo update-alternatives --config g++
```
